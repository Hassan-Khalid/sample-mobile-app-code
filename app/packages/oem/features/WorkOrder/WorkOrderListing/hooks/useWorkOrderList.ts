import { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import WorkOrderListService from '../service/workOrderListService';
import { setMyWorkOrderList, setWorkOrderList } from 'common/redux/slice/workOrderListSlice';
import { bottomTabSelector, myWorkOrderSelector, OfflineModeSelector, workOrderSelector } from 'common/redux/selector';
import { WorkOrderListTypeObject } from 'common/declarations/global';
import { NavigationService } from 'common/services';
import SCREENS from 'common/constants/Screens';


const KEY_DELAY = 1000;

const useWorkOrderList = () => {
  
  const [pageNumber,setPageNumber] = useState(1);
  const [workOrder,setWorkOrder] = useState <WorkOrderListTypeObject[] | [] >([])
  const [myWorkOrder,setMyWorkOrder] = useState <WorkOrderListTypeObject[] | [] >([])
  const isOffline = useSelector(OfflineModeSelector)
  const  bottomTab = useSelector(bottomTabSelector)
  
  const workOrderService = new WorkOrderListService()
  const dispatch = useDispatch();
  const  workOrders  = useSelector(workOrderSelector)
  const  myWorkOrders  = useSelector(myWorkOrderSelector)
  
  const debounce = (func, delay) => {
    let timerId;
    return function (...args) {
      if (timerId) {
        clearTimeout(timerId);
      }
      timerId = setTimeout(() => {
        func(...args);
      }, delay);
    };
  };
  
  const filterOnSearch =(searchKey = '') => {
    if(searchKey === ''){
      setWorkOrder(workOrders)
      setMyWorkOrder(myWorkOrders)
      return
    }
    if(bottomTab === 'MY_WORK_ORDER'){
      let filteredWorkOrders =  myWorkOrders
      .filter((item) =>
        item?.title?.toLowerCase().includes(searchKey.toLowerCase()),
      );
      setMyWorkOrder(filteredWorkOrders)
    }
    else{
      let filteredWorkOrders =  workOrders
      .filter((item) =>
        item?.title?.toLowerCase().includes(searchKey.toLowerCase()),
      );
      setWorkOrder(filteredWorkOrders)
    }
  }
  
  
  const getSearchKey = debounce((val) =>{
    filterOnSearch(val)
  },KEY_DELAY)
  const getWorkOrders = () => {
    const { data:allWorkOrders , loading } = workOrderService.getWorkOrderList();
    if(!loading && !isOffline){
      dispatch(setWorkOrderList({workOrders:allWorkOrders}))
    }
    
  }
  const getMyWorkOrders = () => {
    const { data :allMyWorkOrders , loading:myTicketsLoading} = workOrderService.getMyWorkOrderList();
    if(!myTicketsLoading &&  !isOffline){
      dispatch(setMyWorkOrderList({myWorkOrders:allMyWorkOrders}))
    }
  }
  
  //   const getWorkOrdersWithPage = () =>{
  //   const { data :allMyWorkOrders , loading:myTicketsLoading} = workOrderService.getMyWorkOrderListWithParams(pageNumber);
  //     if(!myTicketsLoading){
  //   }
  // }
  
  const getWorkOrderData = useCallback(() => {
    getWorkOrders()
    getMyWorkOrders()
  }, []);
  
  
  useEffect(() => {
    if(workOrders){
      setWorkOrder(workOrders)
    }
    if(myWorkOrders){
      setMyWorkOrder(myWorkOrders)
    }
  }, [workOrders,myWorkOrders]);
  
  return {
    filterOnSearch,
    getSearchKey,
    getMyWorkOrders,
    getWorkOrders,
    workOrder,
    myWorkOrder,
    getWorkOrderData,
    pageNumber,
    bottomTab,
  };
};
export default useWorkOrderList;
