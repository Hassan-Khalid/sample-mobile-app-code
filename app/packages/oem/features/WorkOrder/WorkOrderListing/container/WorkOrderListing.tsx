import React from 'react';
import { View,FlatList } from 'react-native';
import { ScreenHeaderComponent, Container, BottomTabBar,Text ,WorkOrderItemCard} from 'common/components';
import Util from 'common/util/Util';
import useWorkOrderList from '../hooks/useWorkOrderList';
import CommonConstants from 'common/constants/CommonConstants';
import styles from "../styles/workOrderListing.styles"
import { useSelector } from 'react-redux';
import {  currentUserSelector} from 'common/redux/selector';
import { t } from 'i18next';
import componentBuilder from 'common/builder/ComponentBuilder';

function WorkOrderList() {
  let {getSearchKey, myWorkOrder,getWorkOrders,getMyWorkOrders, workOrder,bottomTab,onWorkOrderPress} = useWorkOrderList();
  getMyWorkOrders()
  getWorkOrders()
  const  user  = useSelector(currentUserSelector)
  const tickets = bottomTab  === 'MY_WORK_ORDER' ?  myWorkOrder: workOrder
  const title = `${t(CommonConstants.tabs[bottomTab])} (${tickets.length})`
  
  const renderEmptyComponent = () => {
    return(
      <View>
        <Text>
          look like you don't have any Work Order......
        </Text>
      </View>
    )
  }
  
  return (
    <>
      <Container {...componentBuilder.createTestAttributes('work-order-listing')}>
        <ScreenHeaderComponent componentId={'screen-header-listing'} title={title} getSearchText={(text)=>getSearchKey(text)}/>
        <View style={styles.container}>
          <FlatList
            ListEmptyComponent={renderEmptyComponent}
            showsVerticalScrollIndicator={false}
            keyExtractor={(item, index) => index.toString() + item._id}
            ItemSeparatorComponent={()=><View style={{margin:5}}></View>}
            contentContainerStyle={{flexGrow:1 ,marginVertical:10,paddingBottom:300,alignItems:'center',width:Util.getWindowWidth()}}
            data={ tickets }
            extraData={[user,workOrder,myWorkOrder]}
            scrollsToTop={true}
            removeClippedSubviews={true}
            updateCellsBatchingPeriod={100}
            // ListFooterComponent={<LoadMoreWorkOrder/>}
            onEndReachedThreshold={0.5}
            renderItem={ (item)=><WorkOrderItemCard { ...item } statuses={user?.oem.statuses} onPress={onWorkOrderPress}/>} />
        </View>
      </Container>
      <BottomTabBar/>
    </>
  );
}
export default WorkOrderList;
