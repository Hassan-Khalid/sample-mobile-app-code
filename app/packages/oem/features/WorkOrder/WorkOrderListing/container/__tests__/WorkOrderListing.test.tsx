import { render } from '@testing-library/react-native';
import WorkOrderListing from '../WorkOrderListing';
import React from 'react';
import useWorkOrderList from '../../hooks/useWorkOrderList';

jest.mock('../../hooks/useWorkOrderList')
// Mock the custom hook

describe('WorkOrderList', () => {
  beforeEach(()=>{
    useWorkOrderList.mockReturnValue({
      getSearchKey: jest.fn(),
      myWorkOrder: [],
      getWorkOrders: jest.fn(),
      getMyWorkOrders: jest.fn(),
      workOrder: [],
      bottomTab: 'ALL_WORK_ORDER',
    })
  })
  
  it('render component', () => {
    //setup
    const wrapper = render(<WorkOrderListing />);
    
    // post-conditions
    expect(wrapper.toJSON).toMatchSnapshot();
  });
  
});
