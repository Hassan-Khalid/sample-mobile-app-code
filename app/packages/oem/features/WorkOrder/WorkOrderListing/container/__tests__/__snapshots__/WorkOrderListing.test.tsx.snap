// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkOrderList render component 1`] = `[Function]`;
